### PR TITLE
[Feature] Add connector components and FitNet

### PR DESCRIPTION
**Pull request by wilxy**
## New Features

1. Add **connector** components. Connector is mainly used for distillation, it usually converts the channel number of input feature to align features of student and teacher. In this PR, `BaseConnector` and three general connectors (`SingleConvConnector`, `ConvBNConnector` and `ConvBNReLUConnector`) are added. 

2. `ConfigurableDistiller` adds two connector related parameters ( `student_connectors` and `teacher_connectors`) and function `build_connectors` to build connectors.

3. Add [**FitNet**](https://arxiv.org/abs/1412.6550) distillation algorithm and its README (where the model and log paths are empty, waiting for the training results of the unified benchmark before the release).

4. Add **L2Loss**. It can be used as the following example.
`from mmrazor.models import L2Loss`
`loss_cfg = dict(loss_weight=10, normalize=False)`
`l2loss = L2Loss(**loss_cfg)`

## Improvement

1. Add `student_trainable` parameter in `SingleTeacherDistill` to control whether the student model is trainable.

2. Add `calculate_student_loss` attribute in `SingleTeacherDistill` to control whether the original task loss to be calculated and used to update student model.

3. Add `test_connectors` and` test_losses` in `tests/test_models`.

### CODE REVIEW

File: mmrazor/models/connectors/base_connector.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+from abc import ABCMeta, abstractmethod
+
+import torch
+import torch.nn as nn
+
+
+class BaseConnector(nn.Module, metaclass=ABCMeta):
+    """Base class of connectors.
+
+    Connector is mainly used for distill, it usually converts the channel
```
Comment (sunnyxiaohu) [inline mmrazor/models/connectors/base_connector.py:11]: distill -> distillation

File: mmrazor/models/connectors/base_connector.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from abc import ABCMeta, abstractmethod
+
+import torch
+import torch.nn as nn
+
+
+class BaseConnector(nn.Module, metaclass=ABCMeta):
```
Comment (sunnyxiaohu) [inline mmrazor/models/connectors/base_connector.py:8]: It's better to inherit from BaseModule?
Then, `self.init_parameters()` could be rewritten by `init_weights()` and `init_cfg`

File: mmrazor/models/connectors/general_connector.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                        m.weight.shape, device=device) / np.sqrt(
+                            k1 * k2 * in_channels) * 1e-4
+                    if hasattr(m, 'bias') and m.bias is not None:
+                        nn.init.zeros_(m.bias)
+                else:
+                    continue
+
+
+@MODELS.register_module()
+class BNConnector(BaseConnector):
```
Comment (sunnyxiaohu) [inline mmrazor/models/connectors/general_connector.py:53]: -> ConvBNConnnector

File: mmrazor/models/connectors/general_connector.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        out_channel (int): The output channel of the connector.
+    """
+
+    def __init__(
+        self,
+        in_channel: int,
+        out_channel: int,
+    ) -> None:
+        super().__init__()
+        self.conv = nn.Conv2d(in_channel, out_channel, kernel_size=1, stride=1)
```
Comment (sunnyxiaohu) [inline mmrazor/models/connectors/general_connector.py:25]: Use `build_conv_layer` to build a conv layer

File: mmrazor/models/connectors/general_connector.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        out_channel (int): The output channels of the connector.
+    """
+
+    def __init__(
+        self,
+        in_channel: int,
+        out_channel: int,
+    ) -> None:
+        super().__init__()
+        self.conv = nn.Conv2d(
```
Comment (sunnyxiaohu) [inline mmrazor/models/connectors/general_connector.py:67]: Use `build_conv_layer` to build a conv layer
`build_norm_layer` to build a norm layer

File: mmrazor/models/connectors/general_connector.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        """Forward computation.
+
+        Args:
+            feature (torch.Tensor): Input feature.
+        """
+        return self.bn(self.conv(feature))
+
+
+@MODELS.register_module()
+class ReLUConnector(BaseConnector):
```
Comment (sunnyxiaohu) [inline mmrazor/models/connectors/general_connector.py:86]: ->ConvBNReLUConnector?

File: mmrazor/models/connectors/general_connector.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        out_channel (int): The output channels of the connector.
+    """
+
+    def __init__(
+        self,
+        in_channel: int,
+        out_channel: int,
+    ) -> None:
+        super().__init__()
+        self.conv = nn.Conv2d(in_channel, out_channel, kernel_size=1)
```
Comment (sunnyxiaohu) [inline mmrazor/models/connectors/general_connector.py:100]: The same

File: mmrazor/models/distillers/configurable_distiller.py [hunk 8]
```diff
@@ -38,6 +38,14 @@ class ConfigurableDistiller(BaseDistiller):
         distill_deliveries (dict, optional): Config for multiple deliveries. A
             distill algorithm may have more than one delivery. Defaults to
             None.
+        student_connectors (dict, optional): Config for multiple connectors. A
```
Comment (sunnyxiaohu) [inline mmrazor/models/distillers/configurable_distiller.py:41]: Explain the key mapping relations between connectors, distill_losses, and loss_forward_mappings in ``Note`` below.

Comment (sunnyxiaohu) [reply mmrazor/models/distillers/configurable_distiller.py:41]: The key mapping relations should be checked. Refers to `_check_loss_forward_mappings`

File: mmrazor/models/distillers/configurable_distiller.py [hunk 9]
```diff
@@ -120,6 +139,22 @@ def prepare_from_teacher(self, model: nn.Module) -> None:
...
+    def build_connectors(
+        self,
+        connectors: Optional[Dict[str, Dict]] = None,
+    ) -> nn.ModuleDict:
+        """Initialize connectors."""
+
+        distill_connecotrs = nn.ModuleDict()
+        if connectors:
+            for loss_name, connector_cfg in connectors.items():
+                assert loss_name in self.distill_losses
```
Comment (sunnyxiaohu) [inline mmrazor/models/distillers/configurable_distiller.py:151]: Add assert_str

File: mmrazor/models/distillers/configurable_distiller.py [hunk 10]
```diff
@@ -120,6 +139,22 @@ def prepare_from_teacher(self, model: nn.Module) -> None:
...
+                assert loss_name not in distill_connecotrs
```
Comment (sunnyxiaohu) [inline mmrazor/models/distillers/configurable_distiller.py:152]: No need?

**Comment by pppppM** [comment]
I have three comments about this PR:
1. `calculate_student_loss ` should be an attribute of the distillation algorithm, not the distiller
2.  The current mapping between connectors and features cannot handle the loss, which has multiple inputs and each input has a different connector.
3. `mmrazor.models.connector` can be moved to `mmrazor.models.architectures`

For comment 2, I think you can refer to the mapping method between recorder, loss, and feature

File: configs/distill/mmcls/fitnet/fitnet_backbone_logits_resnet50_resnet18_8xb32_in1k.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+                in_channel=256,
+                out_channel=1024,
+                norm_cfg=dict(type='BN'))),
+        loss_forward_mappings=dict(
+            loss_s4=dict(
+                s_feature=dict(
+                    from_student=True,
+                    recorder='bb_s4',
+                    record_idx=1,
+                    connector_name='loss_s4_sfeat'),
```
Comment (pppppM) [inline configs/distill/mmcls/fitnet/fitnet_backbone_logits_resnet50_resnet18_8xb32_in1k.py:54]: connector_name -> connector

File: mmrazor/models/distillers/configurable_distiller.py [hunk 12]
```diff
@@ -148,25 +182,30 @@ def get_record(self,
                    recorder: str,
                    from_student: bool,
                    record_idx: int = 0,
-                   data_idx: Optional[int] = None) -> List:
+                   data_idx: Optional[int] = None,
+                   connector_name: Optional[str] = None) -> List:
```
Comment (pppppM) [inline mmrazor/models/distillers/configurable_distiller.py:186]: connector_name -> connector

**Review by pppppM — COMMENTED** [review]
LGTM if the latest comments are fixed.

**Review by pppppM — APPROVED** [review]

**Review by sunnyxiaohu — APPROVED** [review]

**Comment by OpenMMLab-Team** [comment]
Hi @wilxy ！First of all, we want to express our gratitude for your significant PR in the MMRazor project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/UjgXkPWNqA

If you have WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution！❤
